_posts/2016-03-16-AngularTS.md

```diff
@@ -0,0 +1,413 @@
+           });
+   }
+   routes.$inject = ["$routeProvider"];
+
+   angular.module("jworks360")
+       .config(routes);
+}
+{% endhighlight %}
+
+## Services
+When you make a custom service, the code you implement is reusable and can be called in any other angular components including controllers and other services.
+It is important to know that services are singletons, so there will be only one instance of the used service.
+With this in mind we can use the custom service to share data across all components in angular for example communicating with a http service to collect data and share it with any other component by injecting the service.
+
+### RestAngular
+For my project I used an Angular service that simplifies common verb requests with a minimum of client code.
+In my custom services you will see examples of restangular in TypeScript.
+If you like to checkout what the difference is with $resource, you can check this [list](https://github.com/mgonto/restangular#differences-with-resource)
+
+{% highlight javascript %}
+/// <reference path="../../../typings/angularjs/angular.d.ts" />
+/// <reference path="../../../typings/restangular/restangular.d.ts" />
+
+
+module JWorks {
+
+   export interface IEmployeeService {
+       username:string;
+       employee:IEmployee;
+
+       getEmployee(): Employee;
+       setEmployee(employee:IEmployee);
+       getEmployeeByUsername(username:string):ng.IPromise<{}>;
+       getEmployeeByLink(href:string):ng.IPromise<{}>;
+   }
+
+   export class EmployeeService implements IEmployeeService {
+
+       static $inject = ["EmployeeRestangular", "$location"];
+
+       constructor(private employeeRestService:restangular.IService, private $location:ng.ILocationService, private employees:restangular.IElement,public employee:IEmployee,public username:string) {
+            username =window.sessionStorage.getItem("username");
+            this.employees = employeeRestService.all("employees");
+            employees.one(username).get().then((data:any)=> {
+                   this.setEmployee(data);
+               });
+           }
+       }
+              getEmployee():Employee {
+                  return this.employee;
+              }
+       
+              setEmployee(employee:IEmployee) {
+                  this.employee = employee;
+              }
+       
+              getEmployeeByUsername(username:string):ng.IPromise<{}> {
+                  return this.employees.one(username).get()
+       
+              }
+              getEmployeeByLink(href:string):ng.IPromise<{}>{
+                  return this.employees.oneUrl(href).get();
+              }       
+          }
+          angular.module("jworks360")
+              .service("EmployeeService", EmployeeService);       
+       }
+{% endhighlight %}
+
+In the above example, restangular has its own tsd file to reference to.
+For services it is best practice to declare an interface for data typing and get a clear view of the intent.
```

Comment:
Change to:
... it is **a** best practice ...
... data typing and **getting** a clear view ...